Functional programming 👉 partial functions

pmcnr-hx opened the pull request:
Some of us are partial (pun intended) to the functional programming paradigm. This can be a bit confusing for those who have less exposure to the _functional_ way of doing things. We don't claim that _functional_ is better than _imperative_ (although it mostly is) but try instead to clarify the functional alternatives to the more traditional constructs.
- [x] :+1: 
- [x] :+1: :+1:

hx-markterry commented on serverside-javascript-best-practices.md:
````diff
@@ -680,6 +680,65 @@ async.map(['http://foo', 'http://bar'], dbInfo, ...);
+
+  ```javascript
+var subtractFrom5 = function (x) {
+  return subtract(5, x);
````
There is a problem with the indentation here, when the markdown is viewed on github.

hx-markterry commented:
I agree with the above, but it should be moved to the general javascript best practices document as it is not server specific.

pmcnr-hx commented:
@hx-markterry: I've added here because @hxoliverrumbelow mentioned we would be merging the general javascript best practices document with this one. Will we actually be doing that sometime soon or will these always be separate documents and we'll just move part of the server-side best practices that apply generally to the general document?

hx-markterry commented:
I'm sure there will be some server or client specific parts of the javascript best practices, how it is arranged does not really matter as long as it is clear. My main concern here is that the best practice suggested above applies to both and should be under the general document, rather than deferring the merging of practices to someone else or another point in time.

pmcnr-hx commented:
Will move to general @hx-markterry.

pmcnr-hx committed:
Functional programming section with partial.

New section devoted to the functional programming paradigm with a
sub-section clarifying partial function application.

pmcnr-hx commented:
@hx-markterry: moved to general best practices.

hx-markterry commented:
ok with me :monkey:

pmcnr-hx commented:
This has now been rebased on current master.

philmeehan commented:
:+1:
